internal/app/cache/cache_test.go

```diff
@@ -0,0 +1,154 @@
+		Id:      "id_A",
+		Cluster: "cluster_A",
+	},
+	ResourceNames: []string{"resource_A"},
+	TypeUrl:       "typeURL_A",
+	ResponseNonce: "nonce_A",
+}
+
+var testResponse = v2.DiscoveryResponse{
+	VersionInfo: "version_A",
+	Resources:   []*any.Any{},
+	Canary:      false,
+	TypeUrl:     "typeURL_A",
+	Nonce:       "nonce_A",
+	ControlPlane: &core.ControlPlane{
+		Identifier: "identifier_A",
+	},
+}
+
+func TestAddRequestAndFetch(t *testing.T) {
+	cache := NewCache(1, testOnEvict, time.Second*60)
+
+	response, err := cache.Fetch(testKeyA)
+	assert.EqualError(t, err, "no value found for key: key_A")
+	assert.Nil(t, response)
+
+	isStreamOpen, err := cache.AddRequest(testKeyA, testRequest)
+	assert.NoError(t, err)
+	assert.True(t, isStreamOpen)
+
+	response, err = cache.Fetch(testKeyA)
+	assert.NoError(t, err)
+	assert.Nil(t, response)
+}
+
+func TestSetResponseAndFetch(t *testing.T) {
+	cache := NewCache(1, testOnEvict, time.Second*60)
+
+	// Simulate cache miss and setting of new response.
+	response, err := cache.Fetch(testKeyA)
+	assert.EqualError(t, err, "no value found for key: key_A")
+	assert.Nil(t, response)
+
+	requests, err := cache.SetResponse(testKeyA, testResponse)
+	assert.NoError(t, err)
+	assert.Nil(t, requests)
+
+	response, err = cache.Fetch(testKeyA)
+	assert.NoError(t, err)
+	assert.Equal(t, testResponse, *response)
+}
+
+func TestAddRequestAndSetResponse(t *testing.T) {
+	cache := NewCache(1, testOnEvict, time.Second*60)
+
+	isStreamOpen, err := cache.AddRequest(testKeyA, testRequest)
+	assert.NoError(t, err)
+	assert.True(t, isStreamOpen)
+
+	isStreamOpen, err = cache.AddRequest(testKeyA, testRequest)
+	assert.NoError(t, err)
+	assert.True(t, isStreamOpen)
+
+	requests, err := cache.SetResponse(testKeyA, testResponse)
+	assert.NoError(t, err)
+	assert.Equal(t, 2, len(requests))
+	assert.Equal(t, testRequest, *requests[0])
+	assert.Equal(t, testRequest, *requests[1])
+
+	response, err := cache.Fetch(testKeyA)
+	assert.NoError(t, err)
+	assert.Equal(t, testResponse, *response)
+}
+
+func TestMaxEntries(t *testing.T) {
+	cache := NewCache(1, testOnEvict, time.Second*60)
+
+	_, err := cache.SetResponse(testKeyA, testResponse)
+	assert.NoError(t, err)
+
+	response, err := cache.Fetch(testKeyA)
+	assert.NoError(t, err)
+	assert.Equal(t, testResponse, *response)
+
+	_, err = cache.AddRequest(testKeyB, testRequest)
+	assert.NoError(t, err)
+
+	response, err = cache.Fetch(testKeyA)
+	assert.EqualError(t, err, "no value found for key: key_A")
+	assert.Nil(t, response)
+
+	response, err = cache.Fetch(testKeyB)
+	assert.NoError(t, err)
+	assert.Nil(t, response)
+}
+
+func TestTTL_Enabled(t *testing.T) {
+	gomega.RegisterTestingT(t)
+	cache := NewCache(1, testOnEvict, time.Millisecond*10)
+
+	_, err := cache.SetResponse(testKeyA, testResponse)
+	assert.NoError(t, err)
+
+	response, err := cache.Fetch(testKeyA)
+	assert.NoError(t, err)
+	assert.Equal(t, testResponse, *response)
+
+	gomega.Eventually(func() (*v2.DiscoveryResponse, error) {
```

Comment:
since we are handling ttl on our own, can we try to eliminate `Eventually` by passing in a mock time?

Reply:
additional context in https://envoyproxy.slack.com/archives/CTULFMN91/p1584655717039600, but I have replaced the call to Eventually with a short time.Sleep to emulate how other cache libraries test expiration.